### PR TITLE
Update common tests Trainer

### PR DESCRIPTION
**Pull request by SunMarc**
# What does this PR do?

This PR refactor the common tests that we have in Trainer. 

I've mainly did the following: 
- Split the tests that we have in `test_trainer.py` into multiple files. 
- Fix common tests that were failing in the CI

### CODE REVIEW

File: tests/trainer/trainer_test_utils.py [hunk 1]
```diff
@@ -0,0 +1,645 @@
...
+    shift_logits = shift_logits.view(-1, vocab_size)
+    shift_labels = shift_labels.view(-1)
+    # Enable model parallelism
+    shift_labels = shift_labels.to(shift_logits.device)
+    if num_items_in_batch is None or disable_num_items_in_batch:
+        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100, reduction="mean")
+    else:
+        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100, reduction="sum")
+        loss = loss / num_items_in_batch
+    return loss
```
Comment (winglian) [inline tests/trainer/trainer_test_utils.py]: would it make sense to import from https://github.com/huggingface/transformers/blob/main/src/transformers/loss/loss_utils.py#L45-L67

Comment (SunMarc) [reply tests/trainer/trainer_test_utils.py]: yes done !

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1572,7 +1572,9 @@ def _prepare_for_training(self, max_steps, train_dataloader, resume_from_checkpo
         if not delay_optimizer_creation:
             self.create_optimizer()
 
-        model = self._wrap_model(self.model)
+        # Pass `self.model_wrapped` so that `_wrap_model` can detect if the model is already
+        # wrapped (e.g. in DataParallel) on subsequent `train()` calls and avoid double wrapping.
+        model = self._wrap_model(self.model_wrapped)
```
Comment (SunMarc) [inline src/transformers/trainer.py:1577]: regression that we fix here

**Review by winglian — APPROVED** [review]
